### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [17]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
df_purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [18]:
player = df_purchase_data["SN"].unique()

total_players = len(player)

total_players_df = pd.DataFrame({"Total Players": [total_players]})
total_players_df

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [19]:
# Counting total number of unique items
item = df_purchase_data["Item ID"].unique()
item_count = len(item)
print (item_count)

# Calculating average purchase price
avg_price = df_purchase_data["Price"].mean()
print(avg_price)

#Total number of purchases
total_purchases = len(df_purchase_data["Purchase ID"])
print(total_purchases)

# Total revenue
revenue = df_purchase_data["Price"].sum()
print(revenue)

# Formatting and displaying the summary
summary_df = pd.DataFrame({"Unique Items": [item_count], 
                        "Average Price": [avg_price], 
                        "Total Purchases": [total_purchases], 
                        "Total Revenue": [revenue]})

summary_df["Average Price"] = summary_df["Average Price"].map("${:,.2f}".format)
summary_df["Total Purchases"] = summary_df["Total Purchases"].map("${:,}".format)
summary_df["Total Revenue"] = summary_df["Total Revenue"].map("${:,.2f}".format)
summary_df = summary_df.loc[:, ["Unique Items", "Average Price", "Total Purchases", "Total Revenue"]]

summary_df



179
3.0509871794871795
780
2379.77


Unique Items Average Price Total Purchases Total Revenue
0           179         $3.05            $780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [20]:
gender_group = df_purchase_data.groupby("Gender")
count_by_gender_group = gender_group["SN"].nunique()
gender_percent = count_by_gender_group / total_players * 100
gender_demographics = pd.DataFrame({
    "Total Count" : count_by_gender_group,
    "Percentage of Players" : gender_percent
})

print(gender_demographics.to_string(formatters={"Percentage of Players" : "{:,.2f}%".format}))






                       Total Count Percentage of Players
Gender                                                  
Female                          81                14.06%
Male                           484                84.03%
Other / Non-Disclosed           11                 1.91%



## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [21]:
gender_group_total_purchases = gender_group["Purchase ID"].count()
gender_group_avg_price = gender_group["Price"].mean()
gender_group_revenue = gender_group["Price"].sum()
gender_group_avg_total_purchase_person = gender_group_revenue / count_by_gender_group

purchasing_analysis_gender_df = pd.DataFrame({"Purchase Count" : gender_group_total_purchases,
                                              "Average Purchase Price" : gender_group_avg_price,
                                              "Total Purchase Value" : gender_group_revenue,
                                               "Average Total Per Person" : gender_group_avg_total_purchase_person})
print(purchasing_analysis_gender_df.to_string(formatters={"Average Purchase Price" : "${:,.2f}".format,
                                                          "Total Purchase Value" : "${:,.2f}".format,
                                                          "Average Total per Person" : "${:,.2f}".format}))

                       Purchase Count Average Purchase Price Total Purchase Value  Average Total Per Person
Gender                                                                                                     
Female                            113                  $3.20              $361.94                  4.468395
Male                              652                  $3.02            $1,967.64                  4.065372
Other / Non-Disclosed              15                  $3.35               $50.19                  4.562727


## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [22]:
# Establish bins for ages
age_bins = [0, 9, 14, 19, 24, 29, 34, 39, 1000]
age_groups = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

# Categorizing the existing players
df_purchase_data["Age Groups"] = pd.cut(df_purchase_data["Age"], age_bins, labels = age_groups, include_lowest=True)
df_purchase_data

age_group = df_purchase_data.groupby(["Age Groups"])
# Calculating the numbers and percentages by age group
count_by_age_group = age_group["SN"].nunique()
age_groups_percent = count_by_age_group / total_players * 100

# Rounding up the percentage column and display of Age Demographics Table
age_demographics  = pd.DataFrame({"Total Count" : count_by_age_group,
                                    "Percentage of Players" : age_groups_percent})
print(age_demographics.to_string(formatters={"Percentage of Players" : "{:,.2f}%".format}))





            Total Count Percentage of Players
Age Groups                                   
<10                  17                 2.95%
10-14                22                 3.82%
15-19               107                18.58%
20-24               258                44.79%
25-29                77                13.37%
30-34                52                 9.03%
35-39                31                 5.38%
40+                  12                 2.08%


## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [23]:
# Calculating purchase count, average purchase price, average purchase total per person
purchase_count = age_group["Purchase ID"].count()
avg_price = age_group["Price"].mean()
total_purchases = age_group["Price"].sum()
avg_total_purchase_pp = total_purchases / count_by_age_group

# Creating a summary DataFrame
age_purchase_df = pd.DataFrame ({"Purchase Count": purchase_count,
                                 "Average Purchase Price": avg_price,
                                 "Total Purchase Value": total_purchases,
                                  "Average Total Purchase per Person": avg_total_purchase_pp})

# Implementing clear format and displaying the summary DataFrame
age_purchase_df["Average Purchase Price"] = age_purchase_df["Average Purchase Price"].map("${:,.2f}".format)
age_purchase_df["Total Purchase Value"] = age_purchase_df["Total Purchase Value"].map("${:,.2f}".format)
age_purchase_df["Average Total Purchase per Person"] = age_purchase_df["Average Total Purchase per Person"].map("${:,.2f}".format)

age_purchase_df[["Purchase Count", "Average Purchase Price", "Total Purchase Value", "Average Total Purchase per Person"]]



Purchase Count Average Purchase Price Total Purchase Value  \
Age Groups                                                               
<10                     23                  $3.35               $77.13   
10-14                   28                  $2.96               $82.78   
15-19                  136                  $3.04              $412.89   
20-24                  365                  $3.05            $1,114.06   
25-29                  101                  $2.90              $293.00   
30-34                   73                  $2.93              $214.00   
35-39                   41                  $3.60              $147.67   
40+                     13                  $2.94               $38.24   

           Average Total Purchase per Person  
Age Groups                                    
<10                                    $4.54  
10-14                                  $3.76  
15-19                                  $3.86  
20-24                                  $4.32  
25-29                                  $3.81  
30-34                                  $4.12  
35-39                                  $4.76  
40+                                    $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [24]:
top_spenders_group_df = df_purchase_data.groupby(["SN"])

purchase_count = top_spenders_group_df["Purchase ID"].count()
avg_purchase_price = top_spenders_group_df["Price"].mean()
total_purchase_value = top_spenders_group_df["Price"].sum()

# Creating a summary DataFrame
top_spenders_df = pd.DataFrame ({"Purchase Count": purchase_count,
                              "Average Purchase Price": avg_purchase_price,
                              "Total Purchase Value": total_purchase_value})

# Sorting the total purchase value column in descending order
top_spenders_df = top_spenders_df.sort_values(["Total Purchase Value"], ascending=False).head()

# Implementing clear format and displaying the summary DataFrame
top_spenders_df["Average Purchase Price"] = top_spenders_df["Average Purchase Price"].map("${:,.2f}".format)
top_spenders_df["Total Purchase Value"] = top_spenders_df["Total Purchase Value"].map("${:,.2f}".format)

top_spenders_df[["Purchase Count", "Average Purchase Price", "Total Purchase Value"]]


Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [25]:
# Retreiving the Item ID, Item Name, and Item Price columns
popular_items_df = df_purchase_data.loc[:,["Item ID", "Item Name", "Price"]]


# Grouping by Item ID and Item Name
purchase_items_purchase_count = popular_items_df.groupby(["Item ID", "Item Name"]).count()["Price"].rename("Purchase Count")
purchase_items_price = popular_items_df.groupby(["Item ID", "Item Name"]).mean()["Price"]
purchase_items_total_purchase_value = popular_items_df.groupby(["Item ID", "Item Name"]).sum()["Price"].rename("Total Purchase Value")


# Calculating purchase count, item price, and total purchase value
popular_items_purchase_count = popular_items_df["Item ID"].count()
popular_items_price = popular_items_df["Price"].mean()
popular_items_total_purchase_value = popular_items_df["Price"].sum()

# Creating a summary DataFrame
popular_items_df = pd.DataFrame({"Total Purchase Value" : purchase_items_total_purchase_value,
                                 "Item Price" : purchase_items_price,
                                 "Purchase Count" : purchase_items_purchase_count})
item_data_count_sorted = popular_items_df.sort_values("Purchase Count", ascending=False)

item_data_count_sorted.head(5)



Total Purchase Value  \
Item ID Item Name                                                            
92      Final Critic                                                 59.99   
178     Oathbreaker, Last Hope of the Breaking Storm                 50.76   
145     Fiery Glass Crusader                                         41.22   
132     Persuasion                                                   28.99   
108     Extraction, Quickblade Of Trembling Hands                    31.77   

                                                      Item Price  \
Item ID Item Name                                                  
92      Final Critic                                    4.614615   
178     Oathbreaker, Last Hope of the Breaking Storm    4.230000   
145     Fiery Glass Crusader                            4.580000   
132     Persuasion                                      3.221111   
108     Extraction, Quickblade Of Trembling Hands       3.530000   

                                                      Purchase Count  
Item ID Item Name                                                     
92      Final Critic                                              13  
178     Oathbreaker, Last Hope of the Breaking Storm              12  
145     Fiery Glass Crusader                                       9  
132     Persuasion                                                 9  
108     Extraction, Quickblade Of Trembling Hands                  9

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [26]:
# Sorting the above table by total purchase value in descending order
most_profitable_items = popular_items_df.sort_values("Total Purchase Value", ascending=False)

# Implementing clear formating
most_profitable_items["Total Purchase Value"] = most_profitable_items["Total Purchase Value"].map("${:,.2f}".format)
most_profitable_items["Item Price"] = most_profitable_items["Item Price"].map("${:,.2f}".format)

# Displaying a preview of the data frame
most_profitable_items.head(5)


Total Purchase Value  \
Item ID Item Name                                                           
92      Final Critic                                               $59.99   
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76   
82      Nirvana                                                    $44.10   
145     Fiery Glass Crusader                                       $41.22   
103     Singed Scalpel                                             $34.80   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                      Purchase Count  
Item ID Item Name                                                     
92      Final Critic                                              13  
178     Oathbreaker, Last Hope of the Breaking Storm              12  
82      Nirvana                                                    9  
145     Fiery Glass Crusader                                       9  
103     Singed Scalpel                                             8

### Trend Analysis

 -   84% of the players are identified as male with the majority of the players being between the ages of 15 and 34.
 -   Even though  the purchasing data suggest that male players have made more in-game item purchases, female players have higher purchase per person and means they are more likely to purchase more expensive items.
 -   The group of players aged between 20-24 is the group that spends the most money. 